### ISSUE TITLE
[Backend] Write the Token Exchange Contracts

### ISSUE THREAD

**Issue by kamescg** [kamescg/near-swap#2]
# User Story
As the holder of a token I want the ability to easily swap my tokens.  

# Instructions
- Rewrite the Ethereum Uniswap Smart Contracts for NEAR as NearSwap Smart Contracts

# Technical Requirements
The NearSwap smart contracts should provide the same functionality as the Uniswap smart contracts. 

# References
- https://github.com/Uniswap/uniswap-v1

**Comment by dannielboy58** [comment]
What language are we writing the Nearswap contracts in?

**Comment by kamescg** [comment]
What's your preference?

Just has to be compiled down to WASM.

**Comment by dannielboy58** [comment]
Checked out we Assembly... Looks similar to typescript but should be easier

**Comment by kamescg** [comment]
I think you compile down to WebAssembly as opposed to writing in WebAssembly.

For example you can write Javascript, Typescript, Rust, GO, etc... and compile down to WASM.

It's really up to you what language you want to write it in.

**Comment by dannielboy58** [comment]
Got it...i'll rewrite the contracts in solidity

**Comment by kamescg** [comment]
@dannielboy58 Nice! Are you using to solang library to compile from Solidity to WASM?

https://github.com/hyperledger-labs/solang

**Comment by dannielboy58** [comment]
Solang should work... Will check it out once am almost done with the contract

**Comment by dannielboy58** [comment]
quick question tho
`self.name = 0x556e697377617020563100000000000000000000000000000000000000000000
    self.symbol = 0x554e492d56310000000000000000000000000000000000000000000000000000`

this shows the name and symbol of the uniswap token....will nearswap have its own..if yes..how do i get it

**Comment by dannielboy58** [comment]
Also wanted to confirm if its ok to use https://github.com/pipermerriam/ethereum-datetime as a library for implementing the timestamps...couldn't find an easier way to do it in solidity

**Comment by kamescg** [comment]
Are you using the datetime library for the Uniswap block deadline functionality? Or something else?

**Comment by dannielboy58** [comment]
yes...i wanted to but i found another way to do it by using solidity `now` keyword which gets current block timestamp...the deadline will still have to be provided by the user tho...in unix time format

**Comment by dannielboy58** [comment]
Also, this line references ERC20.sol
 `assert self.token.transferFrom(msg.sender, self, token_amount)`

i might have to import ERC20

**Comment by kamescg** [comment]
Does the NEAR protocol have a sense of Unix timestamps?

Why not just use the blocks like the Uniswap contracts?

**Comment by dannielboy58** [comment]
block numbers?

**Comment by kamescg** [comment]
Yes

**Comment by dannielboy58** [comment]
OK..what about the ERC20 dependency?
